packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts

```diff
@@ -62,23 +61,42 @@ import { EpochTracker, FetchType } from "./epochTracker";
 
 /* eslint-disable max-len */
 
-function convertOdspTree(tree: ITree) {
-    const gitTree: resources.ITree = {
-        sha: tree.id,
-        url: "",
-        tree: [],
-    };
-    for (const entry of tree.entries) {
-        gitTree.tree.push({
-            path: entry.path,
-            mode: "",
-            type: entry.type,
-            size: 0,
-            sha: entry.id,
-            url: "",
-        });
+/**
+ * Build a tree hierarchy base on a flat tree
+ *
+ * @param flatTree - a flat tree
+ * @param blobsShaToPathCache - Map with blobs sha as keys and values as path of the blob.
+ * @returns the hierarchical tree
+ */
+function buildHierarchy(
+    flatTree: ITree,
+    blobsShaToPathCache: Map<string, string> = new Map<string, string>()): api.ISnapshotTree {
+    const lookup: { [path: string]: api.ISnapshotTree } = {};
+    const root: api.ISnapshotTree = { blobs: {}, commits: {}, trees: {} };
+    lookup[""] = root;
+
+    for (const entry of flatTree.entries) {
+        const lastIndex = entry.path.lastIndexOf("/");
+        const entryPathDir = entry.path.slice(0, Math.max(0, lastIndex));
+        const entryPathBase = entry.path.slice(lastIndex + 1);
+
+        // The flat output is breadth-first so we can assume we see tree nodes prior to their contents
+        const node = lookup[entryPathDir];
```

Comment:
Probably no change needed here since it's carried over from before, but when reviewing it feels like this assumption is a little out of scope now. So even though it may always be true, within the scope of this function its a little less clear. I guess part of the definition of `contracts.ITree` is that they are sorted this way, which for now happens to be true since they are always created the same way.

Reply:
Changing comment to better reflect reality